amazon-qldb-driver-core/src/driver.rs

```diff
@@ -60,19 +61,32 @@ impl QldbDriverBuilder {
     /// Build a `QldbDriver` using the AWS SDK for Rust.
     ///
     /// `sdk_config` will be used to create a configured instance of the SDK.
-    pub async fn sdk_config(self, sdk_config: Config) -> QldbResult<QldbDriver<QldbSessionSdk>> {
-        let hyper = Client::https();
-        let client = QldbSessionSdk::new(hyper, sdk_config);
+    /// Note that this configuration uses the "sdk config" as opposed to the
+    /// QLDB specific config.
+    pub async fn sdk_config(
+        self,
+        sdk_config: &aws_types::config::Config,
+    ) -> QldbResult<QldbDriver<DynConnector>> {
+        let client = Client::new(sdk_config);
         self.build_with_client(client).await
     }
 
-    pub async fn build_with_client<C>(self, client: C) -> QldbResult<QldbDriver<C>>
+    /// Builds a `QldbDriver` using the AWS SDK for Rust.
+    ///
+    /// Note that `config` is the service-specific (QldbSession) config. For
+    /// shared config, see [`sdk_config`].
+    pub async fn config(self, config: Config) -> QldbResult<QldbDriver<DynConnector>> {
```

Comment:
Why do these functions need to be async?

Reply:
The client has a connection pool and establishing at least 1 connection is part of instantiating the pool. I'm not super wild about this pattern (see some of the discussion in https://github.com/djc/bb8/issues/102). On the other hand, I do like that it moves the error close to the problematic callsite.

One nice thing about Rust coroutines vs Kotlin coroutines is all the await points are obvious:

```rust
    pub async fn config(self, config: Config) -> QldbResult<QldbDriver<DynConnector>> {
        let client = Client::from_conf(config);
        self.build_with_client(client).await
                                                   // ^^^^^
    }
```

```rust
    pub async fn build_with_client<C>(self, client: Client<C>) -> QldbResult<QldbDriver<C>>
    where
        C: SmithyConnector,
    {
        let ledger_name = self
            .ledger_name
            .ok_or(error::usage_error("ledger_name must be initialized"))?;

        let transaction_retry_policy = Arc::new(Mutex::new(self.transaction_retry_policy));

        let session_pool = Pool::builder()
            .test_on_check_out(false)
            .max_lifetime(None)
            .max_size(self.max_concurrent_transactions)
            .connection_timeout(Duration::from_secs(10))
            .error_sink(Box::new(QldbErrorLoggingErrorSink::new()))
            .build(QldbSessionV2Manager::new(client, ledger_name.clone()))
            .await // <-------
        // ^^^^^
            .map_err(|_| error::todo_stable_error_api())?;

        Ok(QldbDriver {
            ledger_name: Arc::new(ledger_name.clone()),
            session_pool: Arc::new(session_pool),
            transaction_retry_policy,
        })
    }
```